From: Don Howard <dhoward@redhat.com>
Date: Tue, 3 May 2011 15:35:09 -0400
Subject: [net] bluetooth: fix sco information leak to userspace
Message-id: <1304436909-8852-1-git-send-email-dhoward@redhat.com>
Patchwork-id: 35743
O-Subject: [RHEL5.7 PATCH] Bluetooth: sco: fix information leak to userspace
Bugzilla: 681311
CVE: CVE-2011-1078
RH-Acked-by: Don Zickus <dzickus@redhat.com>
RH-Acked-by: Dean Nelson <dnelson@redhat.com>
RH-Acked-by: David S. Miller <davem@redhat.com>

Backport from upstream.
Fixes bz 681311, CVE-2011-1078
Brew build: https://brewweb.devel.redhat.com/taskinfo?taskID=3289598
No regressions seen in rhts.

commit c4c896e1471aec3b004a693c689f60be3b17ac86
Author: Vasiliy Kulikov <segoon@openwall.com>
Date:   Mon Feb 14 13:54:26 2011 +0300

    Bluetooth: sco: fix information leak to userspace

    struct sco_conninfo has one padding byte in the end.  Local
    variable cinfo of type sco_conninfo is copied to userspace with
    this uninizialized one byte, leading to old stack contents leak.

    Signed-off-by: Vasiliy Kulikov <segoon@openwall.com>
    Signed-off-by: Gustavo F. Padovan <padovan@profusion.mobi>

Signed-off-by: Don Howard <dhoward@redhat.com>

diff --git a/net/bluetooth/sco.c b/net/bluetooth/sco.c
index 5438014..a224ced 100644
--- a/net/bluetooth/sco.c
+++ b/net/bluetooth/sco.c
@@ -708,6 +708,7 @@ static int sco_sock_getsockopt(struct socket *sock, int level, int optname, char
 			break;
 		}
 
+		memset(&cinfo, 0, sizeof(cinfo));
 		cinfo.hci_handle = sco_pi(sk)->conn->hcon->handle;
 		memcpy(cinfo.dev_class, sco_pi(sk)->conn->hcon->dev_class, 3);
 
